src/libraries/System.Net.Quic/src/System/Net/Quic/QuicStream.cs

```diff
@@ -525,9 +534,13 @@ public void CompleteWrites()
 
     private unsafe int HandleEventStartComplete(ref START_COMPLETE_DATA data)
     {
+        Debug.Assert(_decrementStreamCapacity is not null);
+
         _id = unchecked((long)data.ID);
         if (StatusSucceeded(data.Status))
         {
+            _decrementStreamCapacity(Type);
+
             if (data.PeerAccepted != 0)
             {
                 _startedTcs.TrySetResult();
```

Comment:
Stupid question: why is `_decrementStreamCapacity(Type);` not aligned with `_startedTcs.TrySet...`?

Reply:
It's aligned with MsQuic calculations. And MsQuic decrements right before signaling to us this event. So technically you can do to negative numbers, but all the streams bellow '0' will have to wait PEER_ACCEPT event.

Reply:
So I added a test for this here: https://github.com/dotnet/runtime/pull/101531/files#diff-82858b107709e3aeb48a21a06b9ac8a7898b6493db3637e97930b44ea4161a9fR234-R286
And I had to change the behavior so that `_(b|un)idirectionalStreamCapacity` never goes into negatives. The changes are in this commit: https://github.com/dotnet/runtime/pull/101531/commits/2710d18adc116e2662baa2593508ee787ad6a9d1

What happens is that MsQuic will report zeros in STREAMS_AVAILABLE if the MAX_STREAMS update doesn't move the needle of the current capacity above 0 (i.e. it might have accepted some pending streams to be opened via PEER_ACCEPTED, but did not provide enough capacity to allow opening additional streams). To balance this out, we now ignore all the decrements that would lead bellow zero. Note that the decrements and STREAMS_AVAILABLE happen serially from a single thread. When we get down to zero, we might have some pending streams that are not peer accepted, when we get STREAMS_AVAILABLE with 0, it means we might have gotten some PEER_ACCEPTED, until we get STREAMS_AVAILABLE with positive value, which means that all pending streams were accepted and now we have some capacity to report.

Feel free to think about this more and poke holes into my logic.

Lastly, I hope to change this and offload the counting fully into MsQuic. But as I mentioned above, I don't want to hold this waiting on the next MsQuic release.

Reply:
Do we want to have the numbers reported from the callback always sum up to the MAX_STREAMS value? The current implementation seems to drop some numbers, try following test


```cs

        [Fact]
        public async Task GetStreamCapacity_SumInvariant()
        {
            int maxStreamIndex = 0;
            const int Limit = 5;

            var clientOptions = CreateQuicClientOptions(new IPEndPoint(0, 0));
            clientOptions.StreamCapacityCallback = (connection, args) =>
            {
                Interlocked.Add(ref maxStreamIndex, args.BidirectionalIncrement);
            };

            var listenerOptions = CreateQuicListenerOptions();
            listenerOptions.ConnectionOptionsCallback = (_, _, _) =>
            {
                var options = CreateQuicServerOptions();
                options.MaxInboundBidirectionalStreams = Limit;
                return ValueTask.FromResult(options);
            };

            (QuicConnection clientConnection, QuicConnection serverConnection) = await CreateConnectedQuicConnection(clientOptions, listenerOptions);

            Assert.Equal(5, Limit);

            Queue<(QuicStream client, QuicStream server)> streams = new();

            for (int i = 0; i < Limit; i++)
            {
                QuicStream clientStream = await clientConnection.OpenOutboundStreamAsync(QuicStreamType.Bidirectional);
                await clientStream.WriteAsync(new byte[1]);
                QuicStream serverStream = await serverConnection.AcceptInboundStreamAsync();
                streams.Enqueue((clientStream, serverStream));
            }

            Queue<Task<QuicStream>> tasks = new();
            // enqeue more stream creations
            for (int i = 0; i < Limit; i++)
            {
                var newClientStreamTask = clientConnection.OpenOutboundStreamAsync(QuicStreamType.Bidirectional);
                Assert.False(newClientStreamTask.IsCompleted, "Stream creation should not be completed synchronously");
                tasks.Enqueue(newClientStreamTask.AsTask());
            }

            // dispose streams
            while (streams.Count > 0)
            {
                var (clientStream, serverStream) = streams.Dequeue();
                await clientStream.DisposeAsync();
                await serverStream.DisposeAsync();

                if (tasks.TryDequeue(out var task))
                {
                    clientStream = await task;
                    await clientStream.WriteAsync(new byte[1]);
                    serverStream = await serverConnection.AcceptInboundStreamAsync();
                    streams.Enqueue((clientStream, serverStream));
                }
            }

            // give time to update the count
            await Task.Delay(1000);

            // by now, we opened and closed 2 * Limit, and expect a budget of 'Limit' more
            Assert.Equal(15, maxStreamIndex);
        }
```

on my machine this produces

```
    System.Net.Quic.Tests.QuicConnectionTests.GetStreamCapacity_SumInvariant [FAIL]
      Assert.Equal() Failure: Values differ
      Expected: 15
      Actual:   10
```

Reply:
But I think we can fix this, from what I see in the source code, we can reliably calculate
- total number of created streams (by counting successful STREAM_STARTED events)
- total number of started streams (PeerAccepted in STREAM_STARTED or PEER_ACCEPTED events)

in addition to current counts, we maintain the above counts also for the moment when the last STREAMS_AVAILABLE was received. Since PEER_STARTED will be fired before STREAMS_AVAILABLE, by keeping the delta we can detect exactly how much of the "new" capacity was consumed by pending streams, so we can reliably calculate the total increment received over the wire.

If my explanation was not clear enough, I can have a go at implementing this myself.

Reply:
That's what I'm working on right now. It's the bold edit in my comment above. Basically, counting the number of PEER_ACCEPTED towards to increments.

Reply:
UPDATE: I added your test and also added another test that cancels the pending `OpenOutboundStream`. Which led me to discover that PEER_ACCEPTED counting is not enough. The canceled streams during open do not receive PEER_ACCEPTED as they get SHUTDOWN with IMMEDIATE flag. So to keep the logic simple, I just wait for the capacity to get above 0 and then I report the difference cumulatively.
Example with capacity 3:
- callback with initial capacity 3
- open 3 streams ==> OK
- try to open another 3 ==> pending
- close first 3 ==> no callback yet, but pending 3 streams get accepted
- close 1 stream ==> callback reports 4
- close remaining 2 streams ==> 2x callback with 1
- together callback called with the sum of 6

This will lead to the same callbacks if you cancel the pending streams in step 3 or not.

Note that when we move the increments counting into MsQuic, we will report the increments in step 4 (based on every MAX_STREAMS). 

Also the current logic with held back callback could be improved with the PEER_ACCEPTED, but we'd still need the logic for canceled streams and I feel like the complicated code is not worth it as we:
1. do not go into negatives with multiple connections (so we don't use this)
2. we plan to switch to MsQuic increment counting (which will slightly alter the behavior anyway)

What's unchanged is that we'll cumulatively (eventually) report the same sum.